server/plugin_release.go

```diff
@@ -0,0 +1,398 @@
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: Cfg.GithubAccessToken})
+	tc := oauth2.NewClient(ctx, ts)
+	client = github.NewClient(tc)
+
+	refs, _, err := client.Git.ListRefs(ctx, Cfg.GithubOrg, repo, nil)
+	if err != nil {
+		return err
+	}
+	checkTag := fmt.Sprintf("refs/tags/%s", tag)
+	for _, ref := range refs {
+		if ref.GetRef() == checkTag {
+			return fmt.Errorf("Tag %s already exist. Please use another one", tag)
+		}
+	}
+
+	ref, _, err := client.Git.GetRef(ctx, Cfg.GithubOrg, repo, "heads/master")
+	if err != nil {
+		return err
+	}
+	tagMessage := fmt.Sprintf("Tag %s", tag)
+	tags := &github.Tag{
+		Tag:     github.String(tag),
+		Message: github.String(tagMessage),
+		Object:  ref.Object,
+	}
+
+	_, _, err = client.Git.CreateTag(ctx, Cfg.GithubOrg, repo, tags)
+	if err != nil {
+		return err
+	}
+
+	tagRef := fmt.Sprintf("tags/%s", tag)
+	refTag := &github.Reference{
+		Ref:    github.String(tagRef),
+		Object: ref.Object,
+	}
+	_, _, err = client.Git.CreateRef(ctx, Cfg.GithubOrg, repo, refTag)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getReleaseArtifacts(tag, repo string) error {
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: Cfg.GithubAccessToken})
+	tc := oauth2.NewClient(ctx, ts)
+	client = github.NewClient(tc)
+
+	wait := 2400
+	ctxRelease, cancel := context.WithTimeout(context.Background(), time.Duration(wait)*time.Second)
+	defer cancel()
+
+	assetID, releaseID, err := checkReleaseArtifact(ctxRelease, client, repo, tag)
+	if err != nil {
+		return err
+	}
+
+	assetURL, err := getAssetDownloadURL(ctx, client, repo, assetID)
+	if err != nil {
+		return err
+	}
+
+	filepath, err := downloadAsset(assetURL)
+	if err != nil {
+		return err
+	}
+
+	fileServerPath, err := copyFileToSigningServer(filepath)
+	if err != nil {
+		return err
+	}
+
+	err = signAsset(fileServerPath)
+	if err != nil {
+		return err
+	}
+
+	filename := strings.Split(filepath, "/")
+	baseFilename := filename[len(filename)-1]
+	localSignedFiles, err := copySignedFile(baseFilename)
+	if err != nil {
+		return err
+	}
+
+	err = uploadSignedArtifcatsToGithub(ctx, client, repo, releaseID, localSignedFiles)
+	if err != nil {
+		return err
+	}
+
+	err = uploadSignedArtifcatsToS3(localSignedFiles)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkReleaseArtifact(ctx context.Context, githubClient *github.Client, repo, tag string) (releaseAssetID, releaseID int64, err error) {
+	LogInfo("Checking the if the release asset is available")
+	for {
+		repoReleases, _, err := client.Repositories.ListReleases(ctx, Cfg.GithubOrg, repo, nil)
+		if err != nil {
+			return -1, -1, err
+		}
+		for _, release := range repoReleases {
+			if release.GetTagName() == tag {
+				if release.GetAssetsURL() != "" && len(release.Assets) != 0 {
+					return release.Assets[0].GetID(), release.GetID(), nil
+				}
+				LogInfo("Release found but no assets yet. Still waiting...")
+			}
+			LogInfo("waiting...")
```

Comment:
Why log `waiting...` for each release that doesn't match the tag? Not actually waiting here.